### DISCUSSION TITLE
minor questions

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#22876] [Q&A General] [answered]
Hi colleagues,

I have 2 minor questions regarding the kernel in moose
https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/kernels/PorousFlowAdvectiveFlux.C

1. where is the pressure gradient term in this kernal since I see this kernel is developed for this term with the link https://mooseframework.inl.gov/falcon/source/kernels/PorousFlowAdvectiveFlux.html
2. I always see the residual part in kernal starting with ::computeQpResidual().
However, in this kernal mentioned above, it uses
PorousFlowAdvectiveFlux::mobility
Does this mean that we can change the name after :: ?

Thank you
Jerry

**Comment by Oops-Qiao** [comment]
@GiudGiud :)

**Comment by GiudGiud** [comment, marked as answer]
Hello

The pressure gradient is added by the base class
https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/kernels/PorousFlowDarcyBase.C

We can't change the name after ::, but the computeQpResidual is implemented in the base class, then it may call routines such as mobility() etc that are defined in the derived classes

Guillaume

**Reply by Oops-Qiao** [reply]
> Hello
> 
> The pressure gradient is added by the base class https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/kernels/PorousFlowDarcyBase.C
> 
> We can't change the name after ::, but the computeQpResidual is implemented in the base class, then it may call routines such as mobility() etc that are defined in the derived classes
> 
> Guillaume

Hi Guillaume,

Thanks a lot for your answer. However, if we take look at the base class, line 510 
https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/kernels/PorousFlowDarcyBase.C, there is another definition for mobility. 
PorousFlowDarcyBase::mobility(unsigned nodenum, unsigned phase) const
{
  return _fluid_density_node[nodenum][phase] / _fluid_viscosity[nodenum][phase];
}
So I am wondering how do you identify the mobility used in the base class is from the other rountines. 
like PorousFlowAdvectiveFlux::mobility, line 40. 
https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/kernels/PorousFlowAdvectiveFlux.C#L40

Thank you again.
Jerry

**Reply by GiudGiud** [reply]
The derived class definition always takes precedence over the base class implementation, even in calls to mobility() from routines in the main class.

The only way to access the main class's definition (when it s overriden by a derived class) is to explicitly call
PorousFlowDarcyBase::mobility() instead of mobility()
